### DISCUSSION TITLE
C# LangVersion 10 and older .NET runtimes

### DISCUSSION THREAD

**Discussion by Oberon00** [open-telemetry/opentelemetry-dotnet#2790] [General]
Hello! According to https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version, C# 7.3 is the newest LangVersion that is supported on the current TargetFrameworks (`netstandard2.0;net461`). For example, C# 10 *requires* .NET 6 as executing runtime according to that document.
Now it seems to technically work (you would have noticed otherwise, I assume), but evidently it's an unsupported scenario. Did you ever discuss this?

EDIT: I found some discussion here https://github.com/open-telemetry/opentelemetry-dotnet/pull/380 but it seem to operate under the assumption that C# 8 is supported on older versions of .NET runtimes, which is not the case I think. As I read https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version, it only lists certain features as explanation for *why* C# 8 is not supported on older .NET versions but does not say "if you don't use these, it's supported". For newer C# versions, it does not even list such features, only stating a minimum .NET version.

**Comment by alanwest** [comment]
In my experience, using a language feature not supported by a target framework generates a compile time error. For example, the screenshot here demonstrates that the `netstandard2.0` build would fail if we tried to introduce an interface containing a default implementation which was introduced in C# 8.

<img width="1006" alt="Screen Shot 2022-01-18 at 5 33 14 PM" src="https://user-images.githubusercontent.com/3676547/150046891-3b53be57-bb81-43e9-9a95-edd09160e651.png">

**Comment by Oberon00** [comment]
That may be the case right now, but it is still a completely unsupported "just happens to accidentally work" thing according to the official documentation.

**Reply by alanwest** [reply]
@cijothomas Curious of your thoughts. Are you aware of any other guidance from Microsoft?

I've spot checked a number of the newer language features from C# 8+ and found them to generate compile time errors for features unsupported by our target frameworks.

Though, [the doc](https://docs.microsoft.com/dotnet/csharp/language-reference/configure-language-version) that @Oberon00 references does state that runtime errors are not out of the question:

> Choosing a language version newer than the default can cause hard to diagnose compile-time and runtime errors.

**Reply by cijothomas** [reply]
Thanks for raising this. I never paid close attention until now.  
Not aware of any other guidance, other than the official one shared above.
Will ask around - adding an agenda to next SIG to discuss more.

**Reply by Oberon00** [reply]
Hi @cijothomas! Were you able to gather any insights here? If not, I guess the tendency is to keep using C# 10 or newer anyway?

**Reply by Oberon00** [reply]
Maybe of interest: The System.Diagnostics package itself (and probably other runtime packages) are compiled with at least C# 8 (not sure what the actual version is?) and target down to net461;netstandard2.0 https://github.com/dotnet/runtime/blob/3a25a7f1cc446b60678ed25c9d829420d6321eba/src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj#L6-L7

EDIT: Seems it uses `preview`, i.e. the very very latest.

1. Nothing defined in the csproj https://github.com/dotnet/runtime/blob/v6.0.1/src/libraries/System.Diagnostics.DiagnosticSource/src/System.Diagnostics.DiagnosticSource.csproj
2. MSBuild will search upwards for Directory.Build.props (this is unlimited; i.e. it will only stop as soon as it finds the file or at file system root, not e.g. when it finds a sln file, see https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2022#search-scope)
3. Will find https://github.com/dotnet/runtime/blob/v6.0.1/src/libraries/Directory.Build.props
4. Which imports (https://github.com/dotnet/runtime/blob/v6.0.1/src/libraries/Directory.Build.props#L15) the root https://github.com/dotnet/runtime/blob/v6.0.1/Directory.Build.props 
5. Which defines the LangVersion here https://github.com/dotnet/runtime/blob/v6.0.1/Directory.Build.props#L267-L270

**Reply by cijothomas** [reply]
@Oberon00 Sorry for not updating back here, this was discussed in 1/25/2022 sig meeting, notes below:
.NET runtime repo does similar things (uses latest lang version, but libraries target net461 etc..)
Given that CI tests run on all frameworks (net461+netcore3.1 and higher), we should be okay here.
(Cijo) - check/file a doc issue in the official doc page to make a formal answer on this.

I consulted owners of DiagnosticSource/MS.Ext.Logging and they confirmed that official packages target the latest language, but still target older frameworks. (as you have discovered already)

**Reply by tarekgh** [reply]
The .NET libraries are always built with the latest compiler version even if libraries targeting down-level versions of the framework. Usually, if targeting down-level version of the framework and trying to use a compiler feature needs a newer version of the runtime, you'll get compilation error. 
Example of that, when using the range syntax like `1..10` or using index syntax like `a[^1]`, these will not compile for netstandard 2.0 and will get a compilation error telling the compiler didn't find the Range or Index types.
